core/ir/riscv64/codec.c

```diff
@@ -0,0 +1,1413 @@
+                case 0: return &instr_infos[OP_c_subw];
+                case 1: return &instr_infos[OP_c_addw];
+                // 2 and 3 are reserved.
+                default: return NULL;
+                }
+            default: return NULL;
+            }
+        default: return NULL;
+        }
+    case 5: return &instr_infos[OP_c_j];
+    case 6: return &instr_infos[OP_c_beqz];
+    case 7: return &instr_infos[OP_c_bnez];
+    default: return NULL;
+    };
+}
+
+/* Decode RVC quadrant 2
+ *
+ * The values are derived from table 16.7 in the RISC-V Instruction Set Manual
+ * Volume I: Unprivileged ISA (ver. 20191213).
+ */
+static inline rv_instr_info_t *
+match_op_2(int funct, int bit11to7, int bit6to2, int bit12, bool rv32, bool rv64)
+{
+    switch (funct) {
+    case 0: return &instr_infos[OP_c_slli];
+    case 1: return &instr_infos[OP_c_fldsp];
+    case 2: return &instr_infos[OP_c_lwsp];
+    case 3:
+        if (rv32)
+            return &instr_infos[OP_c_flwsp];
+        else if (rv64)
+            return &instr_infos[OP_c_ldsp];
+        else
+            return NULL;
+    case 4:
+        switch (bit12) {
+        case 0:
+            if (bit6to2 == 0)
+                return &instr_infos[OP_c_jr];
+            else
+                return &instr_infos[OP_c_mv];
+        case 1:
+            if ((bit11to7 == 0) && (bit6to2 == 0))
+                return &instr_infos[OP_c_ebreak];
+            else if (bit6to2 == 0)
+                return &instr_infos[OP_c_jalr];
+            else
+                return &instr_infos[OP_c_add];
+        default: return NULL;
+        }
+    case 5: return &instr_infos[OP_c_fsdsp];
+    case 6: return &instr_infos[OP_c_swsp];
+    case 7:
+        if (rv32)
+            return &instr_infos[OP_c_fswsp];
+        else if (rv64)
+            return &instr_infos[OP_c_sdsp];
+        else
+            return NULL;
+    default: return NULL;
+    }
+}
+
+static rv_instr_info_t *
+get_rvc_instr_info(uint32_t inst, int xlen)
+{
+    /* 0 is an illegal instruction which is often used as a canary. */
+    if (inst == 0)
+        return &instr_infos[OP_unimp];
+
+    int op = GET_FIELD(inst, 1, 0);
+    int funct = GET_FIELD(inst, 15, 13);
+    int bit11to7 = GET_FIELD(inst, 11, 7);
+    int funct2 = GET_FIELD(inst, 11, 10);
+    int bit12 = BIT(inst, 12);
+    int funct3 = GET_FIELD(inst, 6, 5);
+    int bit6to2 = GET_FIELD(inst, 6, 2);
+    bool rv32 = xlen == 32;
+    bool rv64 = xlen == 64;
+    rv_instr_info_t *info = NULL;
+    uint32_t mask, match;
+
+    switch (op) {
+    case 0: info = match_op_0(funct, rv32, rv64); break;
+    case 1: info = match_op_1(funct, funct2, funct3, bit11to7, bit12, rv32, rv64); break;
+    case 2: info = match_op_2(funct, bit11to7, bit6to2, bit12, rv32, rv64); break;
+    }
+
+    if (info == NULL)
+        return NULL;
+    mask = GET_FIELD(info->info.code, 31, 0);
+    match = GET_FIELD(info->info.code, 63, 32);
+
+    if ((inst & mask) != match)
```

Comment:
Maybe it's in one of the other parts, but are there docs on what the `code` field holds?  What is this mask and match?  What do we need to check if we've made it this far and found a table entry?  Is this just a debug sanity check (should it be ifdef DEBUG?) or there are some bits we haven't yet examined and this is a required last step? I can't tell from this code snapshot.  Add a comment here too even if documented somewhere else?

Reply:
There is a comment on line 57 that states what information is put into which part of instr_info_t.
As for the check, I'll verify tomorrow but I vaguely remember there are some holes in the compressed function encoding that we might fall into if say we get an invalid instruction to decode.

Reply:
I've checked the logic again. This should be a sanity check. I'll change it to ASSERT_MESSAGE().

Reply:
Done.